# `NumPy Fundamental Formulas`
---

### Installation : `pip install numpy` in your virtual environment

In [24]:
# pip install numpy

### Importing the NumPy Library

In [25]:
import numpy as np

### `Methods` 

## `1. np.array()`: 
- Creates a NumPy array from Python lists or tuples.
- Creating the Multidimensional Arrays
- Syntax: `np.array(object)`

1D Array or Vector Array

In [26]:
numpy_array = np.array([1,2,3,4,5])
print(numpy_array)
print(type(numpy_array))

# ndarray --> 1d array

[1 2 3 4 5]
<class 'numpy.ndarray'>


2D Array or Matrix Array

In [27]:
# 2D array: matrix array
two_darray = np.array([[1,2,3], [4,5,6]]) # any number row and columen 
print(two_darray)
print(type(two_darray))


[[1 2 3]
 [4 5 6]]
<class 'numpy.ndarray'>


3D Array or Tensors

In [28]:
# 3d Array: tensor array
# z-axis stand for the 3d axis
# x-axis stand for the row
# y-axis stand for the column
three_darray = np.array([[[1,2,3], [4,5,6]],[[8,9,10], [11,23,13]], [[44,55,66], [18,75,100]]])
print(three_darray)

[[[  1   2   3]
  [  4   5   6]]

 [[  8   9  10]
  [ 11  23  13]]

 [[ 44  55  66]
  [ 18  75 100]]]


## `dtype` Parameter 
- which can help us out with deciding about the size of the data we are going to store inside our array

In [29]:
a1 = np.array([1,2,3,4,5], dtype=np.int32)
a2 = np.array([7,8,9,10,11], dtype=np.int64)

print(a1)
print(a2)
# np.float64 or np.float32

[1 2 3 4 5]
[ 7  8  9 10 11]


In [30]:
np.array([1,2,3,4,5], dtype=int)

array([1, 2, 3, 4, 5])

In [31]:
np.array([7,8,9,10,11], dtype=float)

array([ 7.,  8.,  9., 10., 11.])

In [32]:
np.array([7,8,9,10,11,0, -1], dtype=bool)

array([ True,  True,  True,  True,  True, False,  True])

## `2. np.arange()` 
- Creates values within a range.
- Syntax: `np.arange(start,stop, step)` Note:  like `range` function in python

In [33]:
# In Python
empty_list = []
for i in range(16):
    empty_list.append(i)

print(empty_list)

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]


In [34]:
# In Numpy
r = np.arange(16) # start from 0 by default and end on 16
print(r)

r1 = np.arange(1,16) # start from 1 and end 15 
print(r1)

r2 = np.arange(2,21,2) # start with 2 , end to 21 and step is 2
print(r2)

[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15]
[ 1  2  3  4  5  6  7  8  9 10 11 12 13 14 15]
[ 2  4  6  8 10 12 14 16 18 20]


## `3. np.reshape()`
- helps us out with changing the shape of the array 
- Syntax `np.shape(row,column)`
- Remember: rows * columns = Total Element

In [35]:
arange_data = np.arange(1,17) # start, end
reshaped_data = arange_data.reshape(4,4) # row, column
reshaped_data2 = arange_data.reshape(2,8) # row, column
# reshaped_data3 = arange_data.reshape(2,9) # ValueError: cannot reshape array of size 16 into shape (2,9
print(reshaped_data)
print(reshaped_data2)
# print(reshaped_data3)

# Note: row * column = should be equal of the length of the array

[[ 1  2  3  4]
 [ 5  6  7  8]
 [ 9 10 11 12]
 [13 14 15 16]]
[[ 1  2  3  4  5  6  7  8]
 [ 9 10 11 12 13 14 15 16]]


## `4. np.ones()`
- Creates an array filled with ones. ( float value by default )
- Sytax: `np.ones(shape)`

In [36]:
np.ones((4,4))

array([[1., 1., 1., 1.],
       [1., 1., 1., 1.],
       [1., 1., 1., 1.],
       [1., 1., 1., 1.]])

In [37]:
np.ones((4,4), dtype=int)

array([[1, 1, 1, 1],
       [1, 1, 1, 1],
       [1, 1, 1, 1],
       [1, 1, 1, 1]])

---

## `5. np.zeros()`: 
- Creates an array filled with zeros.
- Syntax: np.zeros(shap)

In [38]:
np.zeros((4,4))

array([[0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.]])

In [39]:
np.zeros((4,4), dtype=int)

array([[0, 0, 0, 0],
       [0, 0, 0, 0],
       [0, 0, 0, 0],
       [0, 0, 0, 0]])

### `Difference: ones() vs zeros()`

| `ones()`                               | `zeros()`                     |
| -------------------------------------- | ----------------------------- |
| Filled with 1                          | Filled with 0                 |
| Default float                          | Default float                 |
| Used for multiplicative initialization | Used for empty initialization |


## `6. np.full()`
- Creates an array filled with a specified value
- Syntax: `np.full(shape, value)`

In [40]:
arr = np.full((2, 2), 7)
print(arr)

[[7 7]
 [7 7]]


In [41]:
arr = np.full((5, 4), 1)
print(arr)

[[1 1 1 1]
 [1 1 1 1]
 [1 1 1 1]
 [1 1 1 1]
 [1 1 1 1]]


## `7. np.empty()`
- Creates an array without initializing entries.
- Sytax: `np.empty(shape)`


In [42]:
arr = np.empty((2, 2))
print(arr)

[[3.5e-323 3.5e-323]
 [3.5e-323 3.5e-323]]


### Difference: zeros() vs empty()

| `zeros()`          | `empty()`                  |
| ------------------ | -------------------------- |
| Initializes values | Doesn't initialize         |
| Slightly slower    | Faster                     |
| Safe               | Not safe until overwritten |


## `8. np.identity()`
- help us out with Returns or creating an identity matrix or Square Matrix ( means, rows == column).
- Syntax: np.identity()
- Properties

    - Square matrix
    - Main diagonal = 1
    - Others = 0

- Used heavily in

    - Matrix inverse
    - Linear Algebra
    - Machine Learning

In [47]:
# np.identity((3,4)) # throw error
np.identity(3)

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

## `9. np.eye()` 
- Similar to identity() but much more flexible.

In [50]:
np.eye(3)

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

In [51]:
# can create rectangular matrices
np.eye(3, 5)

array([[1., 0., 0., 0., 0.],
       [0., 1., 0., 0., 0.],
       [0., 0., 1., 0., 0.]])

## `9. np.random.random()`
- it helps us out with generating any random values from 0 to 1 inside our matrix
- Syntax: `np.random.random((row,col))`

In [43]:
np.random.random((3,4)) # row , column

array([[0.02271295, 0.76756487, 0.50932672, 0.82956299],
       [0.38953685, 0.30023801, 0.03454061, 0.68751647],
       [0.97194064, 0.81688733, 0.59376432, 0.82002822]])

## `10. np.linspace()`
- it make sure that every value with the given range is going to have similar distance from each other
- or, Creates evenly spaced values between two limits.
- Syntax: `np.linspace(start, stop, num)`
 

In [48]:
arr = np.linspace(0, 10, 5)
print(arr)

[ 0.   2.5  5.   7.5 10. ]


In [49]:
np.linspace(-10,10,10) # necessary for machine learning

array([-10.        ,  -7.77777778,  -5.55555556,  -3.33333333,
        -1.11111111,   1.11111111,   3.33333333,   5.55555556,
         7.77777778,  10.        ])

---

| Function             | Purpose                                         | Syntax                          | Example                      | Output                | When to Use                 |
| -------------------- | ----------------------------------------------- | ------------------------------- | ---------------------------- | --------------------- | --------------------------- |
| `np.array()`         | Convert existing Python data into a NumPy array | `np.array(object)`              | `np.array([1,2,3])`          | `[1 2 3]`             | You already have data       |
| `np.arange()`        | Generate numbers with a fixed step              | `np.arange(start, stop, step)`  | `np.arange(1,10,2)`          | `[1 3 5 7 9]`         | Sequential numbers          |
| `reshape()`          | Change array dimensions                         | `array.reshape(rows, cols)`     | `np.arange(12).reshape(3,4)` | 3×4 matrix            | Rearrange data              |
| `np.ones()`          | Create array filled with 1s                     | `np.ones(shape)`                | `np.ones((2,3))`             | All ones              | Initialization              |
| `np.zeros()`         | Create array filled with 0s                     | `np.zeros(shape)`               | `np.zeros((2,3))`            | All zeros             | Initialization              |
| `np.empty()`         | Create uninitialized array                      | `np.empty(shape)`               | `np.empty((2,2))`            | Random garbage values | High-performance allocation |
| `np.full()`          | Fill array with one value                       | `np.full(shape, value)`         | `np.full((2,2),7)`           | All 7s                | Constant arrays             |
| `np.identity()`      | Create identity matrix                          | `np.identity(n)`                | `np.identity(3)`             | Diagonal 1s           | Linear Algebra              |
| `np.random.random()` | Random numbers between 0 and 1                  | `np.random.random(shape)`       | `np.random.random((2,2))`    | Random floats         | ML, simulations             |
| `np.linspace()`      | Evenly spaced numbers                           | `np.linspace(start, stop, num)` | `np.linspace(0,10,5)`        | `[0. 2.5 5. 7.5 10.]` | Graphs, ML                  |


---

- array() → Use existing data.
- arange() → Auto-generate numbers with a step.
- linspace() → Generate a fixed count of evenly spaced numbers.
- reshape() → Change the array's shape.
- zeros() → Fill with 0.
- ones() → Fill with 1.
- full() → Fill with any chosen value.
- empty() → Allocate memory without initializing values.
- identity() → Square matrix with diagonal 1s.
- random.random() → Random floats in the interval [0, 1).

---

Yes. Your notes cover the **most common array creation methods**, but if your goal is to build **complete NumPy fundamentals (beginner → intermediate)**, there are several important methods missing.

---

## 1. `np.eye()` ⭐⭐⭐⭐⭐ (Highly Recommended)

Similar to `identity()` but much more flexible.

```python
np.eye(3)
```

Output

```python
[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]
```

Can create rectangular matrices.

```python
np.eye(3, 5)
```

Output

```python
[[1. 0. 0. 0. 0.]
 [0. 1. 0. 0. 0.]
 [0. 0. 1. 0. 0.]]
```

### Difference

| `identity()`       | `eye()`               |
| ------------------ | --------------------- |
| Only square matrix | Square or rectangular |
| Less flexible      | More flexible         |

---

## 2. `np.diag()` ⭐⭐⭐⭐

Creates or extracts a diagonal.

```python
np.diag([1,2,3])
```

Output

```python
[[1 0 0]
 [0 2 0]
 [0 0 3]]
```

Extract diagonal

```python
a = np.array([[1,2],[3,4]])

np.diag(a)
```

Output

```python
[1 4]
```

---

## 3. `np.copy()`

Creates a deep copy.

```python
a = np.array([1,2,3])

b = np.copy(a)
```

Useful when learning **views vs copies**.

---

## 4. `np.asarray()` ⭐⭐⭐⭐

Very common in production code.

```python
lst = [1,2,3]

arr = np.asarray(lst)
```

Difference

```python
np.array()
```

always creates a new array.

```python
np.asarray()
```

avoids copying when the input is already a NumPy array.

---

## 5. `np.copyto()`

Copies values into an existing array.

```python
a = np.zeros(5)

b = np.arange(5)

np.copyto(a,b)
```

---

## 6. `np.fromiter()`

Creates an array from an iterator.

```python
gen = (x*x for x in range(5))

np.fromiter(gen,dtype=int)
```

Output

```python
[0 1 4 9 16]
```

---

## 7. `np.fromfunction()`

Creates an array using a function.

```python
np.fromfunction(lambda i,j: i+j,(3,3))
```

Output

```python
[[0 1 2]
 [1 2 3]
 [2 3 4]]
```

Useful for matrix generation.

---

## 8. `np.meshgrid()` ⭐⭐⭐⭐

Very important in

* Machine Learning
* Computer Vision
* Mathematics
* Plotting

```python
x = np.arange(3)

y = np.arange(3)

X,Y = np.meshgrid(x,y)
```

---

## 9. `np.indices()`

Returns indices of an array.

```python
np.indices((2,3))
```

Useful for advanced indexing.

---

## 10. `np.logspace()` ⭐⭐⭐⭐

Like `linspace()` but logarithmic.

```python
np.logspace(1,3,5)
```

Output

```python
[10.
31.62
100.
316.22
1000.]
```

Very useful in Machine Learning and signal processing.

---

## 11. `np.geomspace()`

Geometric spacing.

```python
np.geomspace(1,1000,4)
```

Output

```python
[1.
10.
100.
1000.]
```

---

## 12. `np.tri()`

Creates a lower triangular matrix.

```python
np.tri(4)
```

Output

```python
[[1. 0. 0. 0.]
 [1. 1. 0. 0.]
 [1. 1. 1. 0.]
 [1. 1. 1. 1.]]
```

---

## 13. `np.triu()` / `np.tril()`

Upper and lower triangle.

```python
np.triu(a)

np.tril(a)
```

Used in linear algebra.

---

## 14. `np.repeat()`

```python
np.repeat([1,2,3],2)
```

Output

```python
[1 1 2 2 3 3]
```

---

## 15. `np.tile()` ⭐⭐⭐⭐

Repeats an entire array.

```python
np.tile([1,2],3)
```

Output

```python
[1 2 1 2 1 2]
```

---

## 16. `np.ravel()` / `flatten()`

Convert multidimensional arrays into 1D.

```python
arr.flatten()

arr.ravel()
```

Difference:

* `flatten()` returns a copy.
* `ravel()` returns a view whenever possible.

---

## 17. `np.squeeze()` ⭐⭐⭐⭐

Removes dimensions of size 1.

```python
np.squeeze(arr)
```

Very common in deep learning workflows.

---

## 18. `np.expand_dims()`

Adds a new axis.

```python
np.expand_dims(arr, axis=0)
```

Useful for preparing data for machine learning models.

---

## Recommended order for a complete NumPy fundamentals course

### Array Creation

* ✅ `np.array()`
* ✅ `np.arange()`
* ✅ `np.linspace()`
* ✅ `np.logspace()`
* ✅ `np.geomspace()`
* ✅ `np.zeros()`
* ✅ `np.ones()`
* ✅ `np.full()`
* ✅ `np.empty()`
* ✅ `np.identity()`
* ✅ `np.eye()`
* ✅ `np.diag()`
* ✅ `np.random.random()`

### Shape Manipulation

* ✅ `reshape()`
* ✅ `flatten()`
* ✅ `ravel()`
* ✅ `squeeze()`
* ✅ `expand_dims()`

### Repetition

* ✅ `repeat()`
* ✅ `tile()`

### Copying

* ✅ `copy()`
* ✅ `asarray()`

### Matrix Utilities

* ✅ `tri()`
* ✅ `triu()`
* ✅ `tril()`

### Advanced Array Generation

* ✅ `fromiter()`
* ✅ `fromfunction()`
* ✅ `meshgrid()`
* ✅ `indices()`

For a **NumPy Fundamentals** course aimed at data science and machine learning, your current list covers about **75–80%** of what beginners need. Adding the methods marked with ⭐⭐⭐⭐ or ⭐⭐⭐⭐⭐ would bring it to a solid **90–95%** coverage before moving on to topics like indexing, slicing, broadcasting, universal functions (ufuncs), aggregation, and linear algebra.
